### DISCUSSION TITLE
How to use multiple databases to read (slave) and write (master)?

### DISCUSSION THREAD

**Discussion by panla** [python-gino/gino#770] [Q&A]
in flask-sqlalchemy, we can

```python
from flask_sqlalchemy import SignallingSession, get_state
from flask_sqlalchemy import SQLAlchemy
from flask_sqlalchemy import orm


class RoutingSession(SignallingSession):
    def get_bind(self, mapper=None, clause=None):
        state = get_state(self.app)

        if self._flushing:
            # write
            return state.db.get_engine(self.app)
        else:
            # read
            return state.db.get_engine(self.app, bind='slave')


class RoutingSQLAlchemy(SQLAlchemy):
    def create_session(self, options):
        return orm.sessionmaker(class_=RoutingSession, db=self, **options)


db = RoutingSQLAlchemy()
```

and how to use multiple databases to read (slave) and write (master)?
Is this question appropriate?

**Comment by wwwjfy** [comment]
There is some info here #209

**Reply by panla** [reply]
> There is some info here #209

thanks! 
when to read, when to write? How to set a flag to distinguishes them?
Use request.method to distinguishes them may be not good.

**Reply by wwwjfy** [reply]
I'm afraid you need to explicitly specify to use master or slave.
Gino currently doesn't have a way to tell and I'm not sure if SQLAlchemy can do that.

**Reply by panla** [reply]
> I'm afraid you need to explicitly specify to use master or slave.
> Gino currently doesn't have a way to tell and I'm not sure if SQLAlchemy can do that.

```python
from gino_starlette import Gino

db = Gino(...)

class ModelMixin(object):
        @classmethod
        async def m_save(cls, **kwargs):
            await db.set_bind(MASTER_DSN)
            return await cls(**kwargs).create()
````

I want to re-package create() and use other db_url when create data, but `await db.set_bind()` costs too much time.
How to speed up or use other function?
